### PR TITLE
Remove explicit generator

### LINKED ISSUE #1488: Refactor `explicit_generator`
Right now, it accepts a `pse_vector` so a single generator can connect to multiple targets on the cell.
I propose that we make it more similar to the other generators, which work on a single target at a time.

```
arb::explicit_generator(arb::pse_vector{{0, 1.0, weight_}, {0, 3.0, weight_}, {1, 2.9, weight_}});
```
would become:
```
{arb::explicit_generator(0, weight, {1.0, 3.0}), 
 arb::explicit_generator(1, weight, {2.9})};
```

Meaning that the explicit generator would have the following signature: 

```
event_generator explicit_generator(const cell_lid_type& target, float weight, std::vector<time_type> times);
```

### PR DESCRIPTION
**Pull request by thorstenhater**
- Remove the old, multi-target `event_generator` class in favour of `schedule_generator(tgt, weight, explicit_schedule)`
- Much simplification ensues, `event_generator` is no longer a type-erasing container, but just what 
   `schedule_generator` was before
- Make the label resolution in generators a bit more eager, no longer at simulation time, 
   but now during setup (bit give a wee bit of perf as well)

Closes #1488

### CODE REVIEW

File: arbor/include/arbor/event_generator.hpp [hunk 1]
```diff
@@ -49,130 +50,38 @@ namespace arb {
...
-};
-
-// Convenience routines for making schedule_generator:
 
 // Generate events with a fixed target and weight according to
 // a provided time schedule.
 
-struct schedule_generator {
-    schedule_generator(cell_local_label_type target, float weight, schedule sched):
+struct event_generator {
```
Comment (brenthuisman) [inline arbor/include/arbor/event_generator.hpp:62]: In the PR message, you wrote the opposite: `event_generator` was suppressed in favor of a simplified `schedule_generator`. Accidental swap or am I missing sth?

Comment (brenthuisman) [reply arbor/include/arbor/event_generator.hpp:62]: Grepping thought the docs finds mentions of `event_generator` that might need updates in:
1. `./cpp/hardware.rst`
2. `./cpp/recipe.rst`

File: test/unit/test_recipe.cpp [hunk 2]
```diff
@@ -201,20 +201,21 @@ TEST(recipe, event_generators) {
...
+        std::vector<arb::event_generator>
+            gens_0 = {arb::explicit_generator({"synapse0"}, 0.1, std::vector<arb::time_type>{1.0}),
+                      arb::explicit_generator({"synapse1"}, 0.1, std::vector<arb::time_type>{2.0})},
+            gens_1 = {arb::explicit_generator({"synapse0"}, 0.1, std::vector<arb::time_type>{1.0})};
 
         auto recipe_0 = custom_recipe({cell_0, cell_1}, {{}, {}}, {{}, {}},  {gens_0, gens_1});
         auto decomp_0 = partition_load_balance(recipe_0, context);
 
-        EXPECT_NO_THROW(simulation(recipe_0, context, decomp_0));
+        EXPECT_NO_THROW(simulation(recipe_0, context, decomp_0).run(1, 0.1));
```
Comment (brenthuisman) [inline test/unit/test_recipe.cpp:213]: Why does it need to run now?

**Review by brenthuisman — CHANGES_REQUESTED** [review]
Simplification: I like it. Some questions:

File: arbor/include/arbor/event_generator.hpp (hunk 1 above)
Comment (thorstenhater) [reply arbor/include/arbor/event_generator.hpp:62]: It's been a small dance
- rm event_generator (because that's the point)
- mv schedule_generator event_generator (since it's now the only kind of generator)
- So, now we still have event_generator, but it's different.

good point about the doc.

Comment (thorstenhater) [reply arbor/include/arbor/event_generator.hpp:62]: Update the docs are fine; just callbacks returning an `event_generator`.

File: test/unit/test_recipe.cpp (hunk 2 above)
Comment (thorstenhater) [reply test/unit/test_recipe.cpp:213]: Actually it always had to. See, the old `event_generator` did up-front target checking and threw an error if any
of them (there could be multiple in one, remember) did not exist. The schedule based one **did not** and did 
defer that to runtime **per event**. 

So, an `event_generator` would have been fine if construction passed, but `schedule_generator` only if simulation
runs. This PR makes the check on schedule eager, removes the need to store the resolver and the runtime checking.
Thus, it's simpler, faster, and safer now.

TL;DR: This requirement for `run` is just paranoia. ;)

**Comment by thorstenhater** [comment]
See above: it's simpler, faster, and safer.

**Comment by thorstenhater** [comment]
Yes, these sections are still empty. Is that a requirement for this PR, though? It does not add anything in this regard nor
change any of the related methods.

I added this to #1242 to remind ourselves.

**Review by brenthuisman — APPROVED** [review]
Agree that adding related documentation would be ideal; no need to spin that out to separate PRs for that.
